inputremapper/injection/macros/macro.py

```diff
@@ -587,56 +587,58 @@ async def task(handler):
 
         self.tasks.append(task)
 
-    def add_if_single(self, then, otherwise, timeout=None):
+    def add_if_single(self, then, _else, timeout=None):
         """If a key was pressed without combining it."""
+        # TODO migrate "otherwise" to "else"
         _type_check(then, [Macro, None], "if_single", 1)
-        _type_check(otherwise, [Macro, None], "if_single", 2)
+        _type_check(_else, [Macro, None], "if_single", 2)
 
         if isinstance(then, Macro):
             self.child_macros.append(then)
-        if isinstance(otherwise, Macro):
-            self.child_macros.append(otherwise)
-
-        resolved_timeout = _resolve(timeout, allowed_types=[int, float, None])
+        if isinstance(_else, Macro):
+            self.child_macros.append(_else)
 
         async def task(handler):
             trigger_code = self._triggering_event.code
 
-            async def listener(event: evdev.InputEvent):
-                """receives every event from consumer_control"""
+            listener_done = asyncio.Event()
+
+            success = False
+
+            async def listener(event):
+                nonlocal success
+
                 if event.type != EV_KEY:
-                    return  # ignore anything that is not a key
+                    # ignore anything that is not a key
+                    return
+
                 if event.code != trigger_code and event.value == 1:
-                    self.context.listeners.remove(listener)
-                    await otherwise.run(handler)
+                    # another key was pressed, trigger else
+                    listener_done.set()
+                    return
 
-            self.context.listeners.add(listener)
-            timed_out = False
-            try:
-                if resolved_timeout is not None:
-                    await asyncio.wait_for(
-                        self._trigger_release_event.wait(), resolved_timeout / 1000
-                    )
-                else:
-                    await self._trigger_release_event.wait()
+                if event.code == trigger_code and event.value == 0:
+                    # the trigger was released
+                    success = True
+                    listener_done.set()
+                    return
```

Comment:
This would not work, for combinations. A combination is released if any of the combination keys is released. But trigger_code only holds the code of the event that completed the combination. And the macro has no knowledge about the combination, not even the macro handler has knowledge about it.

Reply:
On main a mapping of `1 + 2 + 3` -> `h(a)` it will continue running the macro even if 1 and 2 are already released.

Reply:
Not sure what the more intuitive variation is, maybe it should stop even if only `1` or `2` is released.

Reply:
I know, as all other combinations on main do. 
I changed it, since it looked like a limitation of the `Unreleased` mechanic and not like a design decision.

If we want combinations to only release if the key, which completed it, is released I can do that in the `CombinationHandler`. But I think it is more intuitive to release a combination as soon as any key of the combination is released. That is also more in line with the behavior for system native combinations like holding shift.

Reply:
ok, lets do it the way you are suggesting.

Reply:
But I still like the use of the asyncio.Event I think we can make the task much simpler:

```python
async def task(handler):
    trigger_code = self._triggering_event.code
    listener_done = asyncio.Event()

    async def listener(event):
        if event.type != EV_KEY:
            # ignore anything that is not a key
            return

        if event.code != trigger_code and event.value == 1:
            # another key was pressed, trigger else
            listener_done.set()
            return

    self.context.listeners.add(listener)

    resolved_timeout = _resolve(timeout, allowed_types=[int, float, None])
    await asyncio.wait(
        [listener_done.wait(), self._trigger_release_event.wait()],
        timeout=resolved_timeout / 1000 if resolved_timeout else None,
        return_when=FIRST_COMPLETED,
    )

    self.context.listeners.remove(listener)

    if not listener_done.is_set() and self._trigger_release_event.is_set():
        await then.run(handler)  # was trigger release
    else:
        await _else.run(handler)

```